rust/src/traits/samplers/bernoulli/sample_bernoulli_rational.tex

```diff
@@ -1,71 +1,52 @@
-        \item returns \texttt{out} where \texttt{out} is $\top$ with probability \texttt{prob}, otherwise $\bot$.
-    \end{itemize}
-     If \texttt{trials} is set, the implementation's runtime is constant.    
-\end{definition}
-
-\begin{proof} 
-An integer sample is taken uniformly at random from $[0, denom)$, where \texttt{denom} is the denominator of \texttt{prob}.
-The implementation then returns $\top$ if the sample is less than the numerator of \texttt{prob}, and $\bot$ otherwise.
-Since only at most \texttt{numer} outcomes of $\top$ are possible, out of \texttt{denom} possible outcomes,
-the implementation returns $\top$ with probability \texttt{prob}.
-
-The implementation runs in constant-time if \texttt{trials} is set.
-\end{proof}
-
-\end{document}
+\documentclass{article} 
+\input{../../../lib.sty} 
+ 
+\title{\texttt{fn sample\_bernoulli\_rational}} 
+\author{Michael Shoemate} 
+ 
+\begin{document} 
+\maketitle 
+ 
+This document proves that the implementations of \rustdoc{traits/samplers/bernoulli/fn}{sample\_bernoulli\_rational} in \asOfCommit{mod.rs}{f5bb719}  
+satisfies its proof definition. 
+ 
+\subsection{Hoare Triple} 
+\subsubsection*{Preconditions} 
+\textbf{Compiler-verified} 
+\begin{itemize} 
+    \item Variable \texttt{prob} must be of type \texttt{RBig}, a rational of bignums. 
+\end{itemize} 
+ 
+\textbf{User-verified} 
+None 
+ 
+\subsubsection*{Pseudocode} 
+ 
+\lstinputlisting[language=Python,firstline=2,escapechar=|]{./pseudocode/sample_bernoulli_rational.py} 
+ 
+\subsubsection*{Postcondition} 
+\begin{definition} 
+    \label{sample-bernoulli} 
+ 
+    For any setting of the input parameters, 
+    returns an error if there is a lack of system entropy or \texttt{prob} is not in $[0, 1]$, 
+    returns true with probability \texttt{prob}, otherwise returns false. 
+\end{definition} 
+ 
+\begin{proof}  
+At \ref{prob-parts}, \texttt{into\_parts} returns the signed numerator and unsigned nonzero denominator of \texttt{prob}. 
+At \ref{numer-parts}, \texttt{into\_parts} returns the sign and magnitude of \texttt{numer}. 
+ 
+An error is raised if the sign is negative (if \texttt{prob} is negative)  
+or if the numerator is greater than the denominator (if \texttt{prob} is greater than one). 
+ 
+Since \texttt{RBig.into\_parts} never returns a zero denominator (this would be an invalid rational), 
+the precondition of \texttt{sample\_uniform\_ubig\_below} is met, 
+therefore by the postcondition the return value is a uniform sample in $[0, denom)$. 
+ 
+By countable additivity, the probability that the uniform sample is less than the numerator is exactly \texttt{prob}, 
+as there are \texttt{numer} possible disjoint outcomes each with probability $1 / \texttt{denom}$. 
+It has been shown that the function returns true with probability \texttt{prob}, otherwise returns false. 
+\end{proof} 
```

Comment:
This proof looks really good!  I initially reviewed the pdf version, which overlaps a lot with the implementation made for floats (different beasts).

This version is neat and clear!  I have nothing to add (not even nits).

Reply:
If you like the proof, give it an approval! Glad you like it. :D